File: leetcode/833_Find_And_Replace_in_String/833_Find_And_Replace_in_String.cpp
class Solution {
public:
    string findReplaceString(string S, vector<int>& indexes, vector<string>& sources, vector<string>& targets) {
        vector<int> charIndexes = vector<int>(S.length(), -1);
        for (int i = 0; i < indexes.size(); i++) {
            charIndexes[indexes[i]] = i;
        }
        string res;
        for (int i = 0; i < charIndexes.size(); i++) {
            if (charIndexes[i] == -1) res += S[i];
            else if (sources[charIndexes[i]] == S.substr(i, sources[charIndexes[i]].length())) {
                res += targets[charIndexes[i]];
                i += sources[charIndexes[i]].length() - 1;
            } else res += S[i];
        }
        return res;
    }
};

File: leetcode/833_Find_And_Replace_in_String/833_Find_And_Replace_in_String.java
class Solution {
    private class triple {
        public int index;
        public String source;
        public String target;
        public triple(){};
        public triple(int index, String source, String target) {
            this.index = index;
            this.source = source;
            this.target = target;
        }
    }
    public String findReplaceString(String S, int[] indexes, String[] sources, String[] targets) {
        List<triple> operations = new ArrayList<>();
        for (int i = 0; i < indexes.length; i++) {
            operations.add(new triple(indexes[i], sources[i], targets[i]));
        }
        operations.sort((triple l, triple r) -> {return l.index - r.index;});
        StringBuilder sb = new StringBuilder();
        int curIndex = 0;
        for (int i = 0; i < operations.size(); i++) {
            if (curIndex != operations.get(i).index) {
                sb.append(S.substring(curIndex, operations.get(i).index));
                curIndex = operations.get(i).index;
            }
            if (operations.get(i).source.equals(
                S.substring(operations.get(i).index, operations.get(i).index + operations.get(i).source.length()))) {
                sb.append(operations.get(i).target);
                curIndex += operations.get(i).source.length();
            }
        }
        if (curIndex < S.length()) sb.append(S.substring(curIndex));
        return sb.toString();
    }
}
